### PR TITLE
dbt-materialize: refactor custom index materialization

### PR DESCRIPTION
**Pull request by ahelium**
Addresses the materialize repo portion of https://github.com/MaterializeInc/materialize/issues/11538:
- deprecate the custom index materialization
- enable the ability to define indexes when creating materializedviews, views, and sources 

### Testing

- [x] This PR has adequate test coverage / QA involvement has been duly considered.

### CODE REVIEW

**Review by benesch — COMMENTED** [review]
Sweet! One note within about the difference between deprecated and removed.

I'd also recommend beefing up the test a little bit to exercise expression indexes, and also checking that the indexes that get created are what we expect:

```diff
diff --git a/misc/dbt-materialize/test/materialize.dbtspec b/misc/dbt-materialize/test/materialize.dbtspec
index 3f01fbe8f..5abad8b4b 100644
--- a/misc/dbt-materialize/test/materialize.dbtspec
+++ b/misc/dbt-materialize/test/materialize.dbtspec
@@ -72,10 +72,14 @@ sequences:
   test_dbt_custom_materializations:
     project: custom_materializations
     sequence:
+      - type: dbt
+        cmd: seed
       - type: dbt
         cmd: run
       - type: relations_equal
         relations: [test_materialized_view, test_source_index]
+      - type: relations_equal
+        relations: [actual_indexes, expected_indexes]
       # TODO(benesch): figure out how to test that the sink emits the correct
       # data. Ideally we'd just ingest the sink back into Materialize with a
       # source and then use a `relations_equal` step, but that's hard to do
@@ -123,7 +127,7 @@ projects:
         models/test_view_index.sql: |
             {{ config(
                 materialized='view',
-                indexes=[{'columns': ['a']}]
+                indexes=[{'columns': ['a', 'length(a)']}]
             ) }}
 
             SELECT * FROM (VALUES ('chicken', 'pig'), ('cow', 'horse')) _ (a, b)
@@ -152,3 +156,24 @@ projects:
             CREATE SINK {{ mz_generate_name('test_sink') }}
             FROM {{ ref('test_source') }}
             INTO AVRO OCF 'test_sink.ocf'
+
+        models/actual_indexes.sql: |
+            SELECT
+                o.name, ic.index_position, ic.on_position, ic.on_expression
+            FROM mz_indexes i
+            JOIN mz_index_columns ic ON i.id = ic.index_id
+            JOIN mz_objects o ON i.on_id = o.id
+            WHERE i.id LIKE 'u%'
+
+        seeds/expected_indexes.csv: |
+            name,index_position,on_position,on_expression
+            expected_indexes,1,1,
+            expected_indexes,2,2,
+            expected_indexes,3,3,
+            expected_indexes,4,4,
+            test_materialized_view,1,1,
+            test_materialized_view_index,1,1,
+            test_materialized_view_index,1,1,
+            test_source_index,1,1,
+            test_view_index,1,1,
+            test_view_index,2,,pg_catalog.length(a)
```

File: misc/dbt-materialize/dbt/include/materialize/macros/materializations/index.sql [hunk 1]
```diff
@@ -14,28 +14,13 @@
...
-  {% do persist_docs(target_relation, model) %}
-
-  {{ run_hooks(post_hooks, inside_transaction=False) }}
-  {{ run_hooks(post_hooks, inside_transaction=True) }}
-
-  {{ return({'relations': [target_relation]}) }}
+        See: https://materialize.com/docs/sql/create-index for more information about creating indexes
+        with materialize.
+        """
+    )}}
```
[the remark this answers is not in the document]
Comment (morsapaes) [reply misc/dbt-materialize/dbt/include/materialize/macros/materializations/index.sql:25]: Yeah, I think we should leave the code in for a few releases and just put out a deprecation warning in the lines of what is used for `dbt-core` [deprecations](https://github.com/dbt-labs/dbt-core/blob/7f953a6d48a4db254b17fd3929d267e4b6069228/core/dbt/deprecations.py#L9):

```
The custom index materialization is deprecated and will be removed in
a future release of dbt-materialize. Please use the `indexes` config instead:

e.g. {{ config(
    materialized = 'view',
    indexes=[
      {'columns': ['column_a']}
    ]
)}}

Documentation for indexes in Materialize can be found in:
    https://materialize.com/docs/sql/create-index
```

The documentation link should be replaced with our version of [this page](https://docs.getdbt.com/reference/resource-configs/postgres-configs#indexes) once it exists.

File: doc/user/layouts/shortcodes/dbt-materializations.html [hunk 2]
```diff
@@ -4,6 +4,6 @@
 view             | Creates a [view](/sql/create-view).
 materializedview | Creates a [materialized view](/sql/create-materialized-view).
 table            | Creates a [materialized view](/sql/create-materialized-view) (actual table support pending [#5266](https://github.com/MaterializeInc/materialize/issues/5266)).
-index            | Creates an [index](/sql/create-index).
+index            | Deprecated. Create a materializedview, view, or source with an [index](/sql/create-index) instead.
```
Comment (morsapaes) [inline doc/user/layouts/shortcodes/dbt-materializations.html:7]:
```suggestion
index            | (Deprecated) Creates an index.
```

File: misc/dbt-materialize/README.md [hunk 3]
```diff
@@ -53,7 +53,7 @@ Type               | Supported? | Details
 `view`             | YES        | Creates a [view].
 `materializedview` | YES        | Creates a [materialized view].
 `table`            | YES        | Creates a [materialized view]. (Actual table support pending [#5266].)
-`index`            | YES        | Creates an [index].
+`index`            | NO         | Deprecated. Create a materializedview, view, or source with an index instead.
```
Comment (morsapaes) [inline misc/dbt-materialize/README.md:56]:
```suggestion
`index`            | YES         | (Deprecated) Creates an index. Use the `indexes` config to create indexes on `materializedview`, `view` or `source` relations instead.
```

File: misc/dbt-materialize/test/materialize.dbtspec [hunk 4]
```diff
@@ -75,7 +75,7 @@ sequences:
       - type: dbt
         cmd: run
       - type: relations_equal
-        relations: [test_materialized_view, test_source]
+        relations: [test_materialized_view, test_source_index]
```
Comment (morsapaes) [inline misc/dbt-materialize/test/materialize.dbtspec:78]:
```suggestion
        relations: [test_materialized_view, test_view_index, test_source_index]
```

**Comment by morsapaes** [comment]
One index-related thing that annoys me is using `CASCADE` instead of the default `RESTRICT` [syntax](https://materialize.com/docs/sql/drop-index/#syntax) in the `materialize__drop_index(index_name)` macro:

```
{% macro materialize__drop_index(index_name) -%}
  {% call statement('drop_index') -%}
    drop index if exists {{ index_name }} restrict
  {%- endcall %}
{% endmacro %}
```

It's totally irrelevant in our case and probably an equally annoying nit to point out from my side, since indexes can't have dependent objects and all variations of the syntax produce the same outcome, but we could patch it just...because?

File: misc/dbt-materialize/dbt/include/materialize/macros/materializations/index.sql (hunk 1 above)
[the remark this answers is not in the document]
Comment (ahelium) [reply misc/dbt-materialize/dbt/include/materialize/macros/materializations/index.sql:30]: Sounds great, thank you.

File: misc/dbt-materialize/setup.py [hunk 5]
```diff
@@ -41,7 +41,7 @@
             "include/materialize/macros/**/*.sql",
         ]
     },
-    install_requires=["dbt-postgres==1.0.4"],
+    install_requires=["dbt-postgres~=1.0.0"],
```
Comment (morsapaes) [inline misc/dbt-materialize/setup.py:44]: Thanks for sneaking this in!

File: misc/dbt-materialize/test/materialize.dbtspec [hunk 6]
```diff
@@ -123,7 +127,7 @@ projects:
         models/test_view_index.sql: |
             {{ config(
                 materialized='view',
-                indexes=[{'columns': ['a']}]
+                indexes=[{'columns': ['a', 'length(a)']}]
```
Comment (morsapaes) [inline misc/dbt-materialize/test/materialize.dbtspec:130]: 👍

File: misc/dbt-materialize/setup.py (hunk 5 above)
Comment (morsapaes) [reply misc/dbt-materialize/setup.py:44]: (Assuming we all agree on #11417.)

**Comment by ahelium** [comment]
Thanks for your help @morsapaes @benesch! Important to note that the index testing may fail later on depending on the results of https://github.com/MaterializeInc/materialize/issues/11630.

**Comment by ahelium** [comment]
How should we handle release notes for changes that have not yet been released? A `dbt-release` tag? Seems like including them in the changelog might muck with dependabot.

**Review by benesch — APPROVED** [review]

**Comment by benesch** [comment]
> Thanks for your help @morsapaes @benesch! Important to note that the index testing may fail later on depending on the results of https://github.com/MaterializeInc/materialize/issues/11630.

Yup, all good. Super easy to update the test when we fix the bug.

> How should we handle release notes for changes that have not yet been released? A dbt-release tag? Seems like including them in the changelog might muck with dependabot.

Just drop them in a new "Unreleased" heading! That's what we've done in the past: https://github.com/MaterializeInc/materialize/blob/c770e89d42ffdd3b0299ef0e3601866fae979afa/misc/dbt-materialize/CHANGELOG.md#unreleased

**Pull request merged**